src/libraries/System.Private.CoreLib/src/System/Text/ASCIIUtility.cs

```diff
@@ -1594,6 +1596,101 @@ private static unsafe nuint NarrowUtf16ToAscii_Sse2(char* pUtf16Buffer, byte* pA
             goto Finish;
         }
 
+        private static unsafe nuint NarrowUtf16ToAscii_AdvSimd(char* pUtf16Buffer, byte* pAsciiBuffer, nuint elementCount)
+        {
+            // This meethod follows similar logic as the NarrowUtf16ToAscii_Sse2 method above.
+
+            uint SizeOfVector128 = (uint)Unsafe.SizeOf<Vector128<byte>>();
+            nuint MaskOfAllBitsInVector128 = (nuint)(SizeOfVector128 - 1);
+
+            Debug.Assert(AdvSimd.Arm64.IsSupported);
+            Debug.Assert(BitConverter.IsLittleEndian);
+            Debug.Assert(elementCount >= 2 * SizeOfVector128);
+
+            // Process the first chunk of 16 bytes
+
+            ref ushort utf16Buffer = ref *(ushort*)pUtf16Buffer;
+            ref byte asciiBuffer = ref *pAsciiBuffer;
+            Vector128<ushort> utf16VectorFirst = Vector128.LoadUnsafe(ref utf16Buffer);
+            Vector128<ushort> maxChars = AdvSimd.Arm64.MaxPairwise(utf16VectorFirst, utf16VectorFirst);
+
+            if ((maxChars.AsUInt64().ToScalar() & 0xFF80FF80FF80FF80) != 0)
+            {
+                return 0; // found non-ASCII char
+            }
+            Vector128<byte> tempRegV = AdvSimd.Arm64.UnzipEven(utf16VectorFirst.AsByte(), utf16VectorFirst.AsByte());
```

Comment:
This can be `Vector128.Shuffle(utf16VectorFirst.AsUInt64(), Vector128.Create(0x0E0C_0A08_0604_0200))`

It might be worth having explicit `Vector128.Unzip` and `Vector128.Zip` APIs in the future, we could even have internally only ones for "right now" to make this more readable.

Reply:
> This can be `Vector128.Shuffle(utf16VectorFirst.AsUInt64(), Vector128.Create(0x0E0C_0A08_0604_0200))`

@tannergooding  - Is this optimal than just `AdvSimd.Arm64.UnzipEven(vector.AsByte(), vector.AsByte());`?

Reply:
[Assembly](https://gist.github.com/SwapnilGaikwad/e0132037e6500af9d8d19f3bebcb6e14#file-sse2_only_vectorcontainsnonasciichar-txt-L83-L85) shows that the duplicate computation of `moveMask` cannot be avoided, not as I expected.
I agree that we shouldn't make existing things slower but still think that we should take this penalty because:
- It's a pentaly of a few cycles in the error case when a non-ASCII char is found in a given chunk on the systems pre-SSE4.1
- It will make it easier to move to the vector API when we can implement `VectorContainsNonAsciiChar` purely using the API
- This patch adds similar pentaly of a few cycles on AArch64 while detecting presence of non-ASCII char in the first or second chunk. [There](https://gist.github.com/SwapnilGaikwad/f5533c5e57e0a277ef26305e716c730d#file-aarch64_less_args_no_debug_asm_dump-txt-L119-L120) we repeate pairwise-max and SIMD to GPR transfer for unified AArch64 and x86 version. We would need more platform specific changes to the loop for reclaiming those cycles, e.g. [here](https://github.com/dotnet/runtime/pull/70080/commits/6ac5e90a8142faef700efbab714250db7ac443ff#diff-66bbe89271f826c9232bd146abb678844754515dc027f70ad0ce36f751da46ebR1661-R1667).

If you want to avoid the additional mask computation then I can [undo refactoring](https://gist.github.com/SwapnilGaikwad/932ff832b342f9c4d9437d71ec21a62b).

Reply:
An unzip instruction has lower latency compared to its shuffle equivalent ([load + table lookup](https://gist.github.com/SwapnilGaikwad/f5533c5e57e0a277ef26305e716c730d#file-aarch64_less_args_no_debug_asm_dump-txt-L60-L61)). However, we use the vector API version so that it will be quicker to move to the purely vector API based intrinsic in the future. Would you prefer `AdvSimd.Arm64.UnzipEven()` over `Vector128.Shuffle()`?

Probably, it would also be useful to enable the `Vector128.Shuffle()` for AArch64 detect the above pattern and emits unzip instead of a table lookup.

Reply:
The unzip approach seems about 2% to 3% faster.
```
|   Method |        Job |                                                                                             Toolchain | size | encName |        Mean |    Error |   StdDev |      Median |         Min |         Max | Ratio | MannWhitney(2%) |  Gen 0 | Allocated | Alloc Ratio |
|--------- |----------- |------------------------------------------------------------------------------------------------------ |----- |-------- |------------:|---------:|---------:|------------:|------------:|------------:|------:|---------------- |-------:|----------:|------------:|
| GetBytes | Job-BHXFXY | /with_table_lookup/bin/testhost/net7.0-Linux-Release-arm64/shared/Microsoft.NETCore.App/7.0.0/corerun |   16 |   ascii |    54.96 ns | 0.031 ns | 0.029 ns |    54.96 ns |    54.92 ns |    55.02 ns |  1.00 |            Base | 0.0763 |      40 B |        1.00 |
| GetBytes | Job-UFFZXG |    /use-unzip-even/bin/testhost/net7.0-Linux-Release-arm64/shared/Microsoft.NETCore.App/7.0.0/corerun |   16 |   ascii |    53.93 ns | 0.075 ns | 0.063 ns |    53.90 ns |    53.86 ns |    54.08 ns |  0.98 |            Same | 0.0764 |      40 B |        1.00 |
|          |            |                                                                                                       |      |         |             |          |          |             |             |             |       |                 |        |           |             |
| GetBytes | Job-BHXFXY | /with_table_lookup/bin/testhost/net7.0-Linux-Release-arm64/shared/Microsoft.NETCore.App/7.0.0/corerun |  512 |   ascii |   307.55 ns | 1.985 ns | 1.856 ns |   308.09 ns |   301.89 ns |   310.13 ns |  1.00 |            Base | 1.0244 |     536 B |        1.00 |
| GetBytes | Job-UFFZXG |    /use-unzip-even/bin/testhost/net7.0-Linux-Release-arm64/shared/Microsoft.NETCore.App/7.0.0/corerun |  512 |   ascii |   302.53 ns | 0.316 ns | 0.280 ns |   302.59 ns |   301.95 ns |   302.94 ns |  0.98 |            Same | 1.0238 |     536 B |        1.00 |
|          |            |                                                                                                       |      |         |             |          |          |             |             |             |       |                 |        |           |             |
| GetBytes | Job-BHXFXY | /with_table_lookup/bin/testhost/net7.0-Linux-Release-arm64/shared/Microsoft.NETCore.App/7.0.0/corerun | 2048 |   ascii | 1,145.64 ns | 3.919 ns | 3.474 ns | 1,144.84 ns | 1,139.57 ns | 1,150.59 ns |  1.00 |            Base | 3.9516 |    2072 B |        1.00 |
| GetBytes | Job-UFFZXG |    /use-unzip-even/bin/testhost/net7.0-Linux-Release-arm64/shared/Microsoft.NETCore.App/7.0.0/corerun | 2048 |   ascii | 1,113.24 ns | 1.497 ns | 1.327 ns | 1,113.09 ns | 1,111.12 ns | 1,116.08 ns |  0.97 |          Faster | 3.9485 |    2072 B |        1.00 |
```

Reply:
If that's the case, I would just delete the `ExtractAsciiVector(vector)` override and keep the `ExtractAsciiVector(vectorFirst, vectorSecond)` and use the `UnzipEven()` unless @tannergooding has strong reason to not do it.

Reply:
Done